internal/pkg/scaffold/olm-catalog/descriptor.go

```diff
@@ -0,0 +1,488 @@
+			if err != nil {
+				return fmt.Errorf("error unquoting %s: %v", val, err)
+			}
+			d.xdesc = strings.Split(xdStr, ",")
+		default:
+			return fmt.Errorf(`unsupported descriptor path element "%s"`, pathElems[1])
+		}
+	default:
+		return fmt.Errorf(`unsupported descriptor path "%s"`, annotations.JoinPath(pathElems...))
+	}
+	return nil
+}
+
+func parseResource(rStr string) (r olmapiv1alpha1.APIResourceReference, err error) {
+	rStr, err = strconv.Unquote(rStr)
+	if err != nil {
+		return r, err
+	}
+	rSplit := strings.SplitN(rStr, ",", 3)
+	if len(rSplit) < 2 {
+		return r, fmt.Errorf("resource string %s did not have at least a kind and a version", rStr)
+	}
+	r.Kind, r.Version = rSplit[0], rSplit[1]
+	if len(rSplit) == 3 {
+		r.Name, err = strconv.Unquote(rSplit[2])
+		if err != nil {
+			return r, err
+		}
+	}
+	return r, nil
+}
+
+func setDescriptorDefaultsIfEmpty(d *descriptor) {
+	if d.description == "" {
+		d.description = parseDescription(d.member.CommentLines)
+	}
+	if d.path == "" {
+		d.path = parsePathFromJSONTags(d.member.Tags)
+	}
+	if d.displayName == "" {
+		d.displayName = getDisplayName(d.member.Name)
+	}
+	switch d.descType {
+	case typeSpec:
+		d.xdesc = getSpecXDescriptorsByPath(d.xdesc, d.path)
+	case typeStatus:
+		d.xdesc = getStatusXDescriptorsByPath(d.xdesc, d.path)
+	}
+}
+
+func getTypeName(t *types.Type) string {
+	return getUnderlyingType(t).Name.String()
+}
+
+func typeNamesEqual(t1, t2 *types.Type) bool {
+	return getTypeName(t1) == getTypeName(t2)
+}
+
+func mergeChildDescriptorPaths(specType, statusType *types.Type, descriptors []descriptor) (newDescs []descriptor) {
+	descMap := map[string][]descriptor{}
+	for _, d := range descriptors {
+		n := getTypeName(d.member.Type)
+		descMap[n] = append(descMap[n], d)
+	}
+	bfsJoinDescriptorPaths(specType, typeSpec, descMap)
+	bfsJoinDescriptorPaths(statusType, typeStatus, descMap)
+	for _, ds := range descMap {
+		for _, d := range ds {
+			newDescs = append(newDescs, d)
+		}
+	}
+	return newDescs
+}
+
+func getUnderlyingType(t *types.Type) *types.Type {
+	switch t.Kind {
+	case types.Map, types.Slice, types.Pointer, types.Chan:
+		t = t.Elem
+	case types.Alias, types.DeclarationOf:
+		t = t.Underlying
+	}
+	return t
+}
+
+type memberNode struct {
+	types.Member
+	parentNode *memberNode
+}
+
+type descNodeMapping struct {
+	parentNode *memberNode
+	descriptor descriptor
+}
+
+func bfsJoinDescriptorPaths(parentType *types.Type, pt descriptorType, descMap map[string][]descriptor) {
+	nextMembers, leaves := []*memberNode{}, []descNodeMapping{}
+	for _, m := range parentType.Members {
+		nextMembers = append(nextMembers, &memberNode{m, nil})
+	}
+	level, lenNextMembers := 0, len(nextMembers)
+	// BFS up to 5 levels.
+	for len(nextMembers) > 0 && level < 5 {
+		for _, m := range nextMembers {
+			t := getUnderlyingType(m.Type)
+			for _, mm := range t.Members {
+				node := memberNode{mm, m}
+				nextMembers = append(nextMembers, &node)
+				mmn := getTypeName(mm.Type)
```

Comment:
We have to check that both the parent types and member types/names are equal before adding a leaf, and in order to do that we must loop through all fields in a parent `struct` type, not just all members in `nextMembers`. We could find an incorrect leaf if we only checked member type/name equality (which is what we'd be doing in the above suggestion) since different structs can have the same field signatures.

I'll add a bunch of comments since this isn't clear at all.